build/source/engine/run_oneGRU.f90

```diff
@@ -0,0 +1,375 @@
+ type(gru_hru_int)       , intent(in)    :: typeStruct             ! x%gru(:)%hru(:)%var(:)     -- local classification of soil veg etc. for each HRU
+ type(gru_hru_double)    , intent(in)    :: attrStruct             ! x%gru(:)%hru(:)%var(:)     -- local attributes for each HRU
+ type(gru_hru_doubleVec) , intent(in)    :: mparStruct             ! x%gru(:)%hru(:)%var(:)%dat -- local (HRU) model parameters
+ ! data structures (input-output)
+ type(gru_hru_intVec)    , intent(inout) :: indxStruct             ! x%gru(:)%hru(:)%var(:)%dat -- model indices
+ type(gru_hru_double)    , intent(inout) :: forcStruct             ! x%gru(:)%hru(:)%var(:)     -- model forcing data
+ type(gru_hru_doubleVec) , intent(inout) :: progStruct             ! x%gru(:)%hru(:)%var(:)%dat -- model prognostic (state) variables
+ type(gru_hru_doubleVec) , intent(inout) :: diagStruct             ! x%gru(:)%hru(:)%var(:)%dat -- model diagnostic variables
+ type(gru_hru_doubleVec) , intent(inout) :: fluxStruct             ! x%gru(:)%hru(:)%var(:)%dat -- model fluxes
+ type(gru_doubleVec)     , intent(inout) :: bvarStruct             ! x%gru(:)%var(:)%dat        -- basin-average variables
+ ! error control
+ integer(i4b)            , intent(out)   :: err                    ! error code
+ character(*)            , intent(out)   :: message                ! error message
+
+ ! ----- define local variables ------------------------------------------------------------------------------------------
+
+ ! general local variables
+ character(len=256)                      :: cmessage               ! error message
+ integer(i4b)                            :: iHRU                   ! HRU index
+ ! HRU initialization
+ real(dp)                                :: fracHRU                ! fractional area of a given HRU (-)
+ logical(lgt)            , parameter     :: overwriteRSMIN=.false. ! flag to overwrite RSMIN
+ integer(i4b)            , parameter     :: maxSoilLayers=10       ! Maximum Number of Soil Layers
+ real(dp)                , allocatable   :: zSoilReverseSign(:)    ! height at bottom of each soil layer, negative downwards (m)
+ ! run the model
+ integer(i4b),parameter                  :: no=0                   ! .false.
+ integer(i4b),parameter                  :: yes=1                  ! .true.
+ logical(lgt)                            :: computeVegFluxFlag     ! flag to indicate if we are computing fluxes over vegetation (.false. means veg is buried with snow)
+ ! compute fluxes across HRUs
+ integer(i4b)                            :: jHRU,kHRU              ! index of the hydrologic response unit 
+
+ ! initialize error control
+ err=0; message='run_oneGRU/'
+ 
+ ! ----- basin initialization --------------------------------------------------------------------------------------------
+
+ ! initialize runoff variables
+ bvarStruct%gru(iGRU)%var(iLookBVAR%basin__SurfaceRunoff)%dat(1)    = 0._dp  ! surface runoff (m s-1)
```

Comment:
Looking through this quickly, I tend to agree with @bartnijssen All of these structs should be indexed at (iGRU) before passing them in, then you don't need to pass iGRU at all.  I don't think there will be any performance penalty for this as you aren't doing a strided array access or other funky indexing.  Then you can (mostly) just remove all of the `%gru(iGRU)` from this code.  This has the added benefit of making the subroutine a little easier to read.  It is a little more work upfront, but it shouldn't be too much work while you have it all in your mind anyway.

Reply:
Another benefit of making this change is it will make it much easier to adapt to MPI if/when that comes around.

Reply:
You would replace all the terms that look like
```Fortran
forcStruct%gru(iGRU)%hru(iHRU)%var,& ! vector of model forcing data
attrStruct%gru(iGRU)%hru(iHRU)%var,& ! vector of model attributes
mparStruct%gru(iGRU)%hru(iHRU),    & ! vector of model parameters
progStruct%gru(iGRU)%hru(iHRU),    & ! data structure of model prognostic variables
diagStruct%gru(iGRU)%hru(iHRU),    & ! data structure of model diagnostic variables
fluxStruct%gru(iGRU)%hru(iHRU),    & ! data structure of model fluxes
...
```
by something like (may not be the right syntax)
```Fortran
forcGRU%hru(iHRU)%var,& ! vector of model forcing data
attrGRU%hru(iHRU)%var,& ! vector of model attributes
mparGRU%hru(iHRU),    & ! vector of model parameters
progGRU%hru(iHRU),    & ! data structure of model prognostic variables
diagGRU%hru(iHRU),    & ! data structure of model diagnostic variables
fluxGRU%hru(iHRU),    & ! data structure of model fluxes
...
```
The only reason to pass in `iGRU` would some logging or error reporting

Reply:
Perhaps you can go even one further `forcHRU(iHRU)%var` .... (this is where my fortran skillz are lacking)

Reply:
yes, I think if the only part of forcGRU accessed is %hru, then you should be able to pass that in and index it as you are indicating. 

This sort of thing could actually lead to a performance benefit as the compiler will know what there is no way you would access anything else internally, and it might be able to avoid a few dereferences.

Reply:
I think that would be good, because it may provide advantages down the line. I was discussing this with @jhamman and he brought up that the current implementation will work with shared memory (`openMP`), but that it won't necessarily (at all?) work with distributed memory (`MPI`). Since we want to move down that road in the near future, it is probably worth it to do the work now. Which makes me wonder how LIS implemented the calls to the individual HRUs or GRUs?

Reply:
Since you want to loop over the HRUs, I think you'd be passing something like 
```
forcStruct%gru(iGRU)
```
and then refer to it inside the function as 
```
forcHRU(iHRU)%var
```
It probably isn't that hard to implement once you have the right pattern.

Reply:
Correct. I just need to identify the correct data type.

Reply:
While you are at it, would it make sense to provide one more level and have run_one_gru just have a loop over hrus and inside that loop call run_one_hru?  That could provide flexibility such that in the future you could have a routine that called run_one_hru directly (e.g. for point simulations).  It is always a good thing to separate the logic into smaller pieces like that too.

Reply:
Yes, the call to `coupled_em` runs one HRU. It will make sense to change the name of this routine (this is all that is required).

Reply:
I wouldn't do that, coupled_em does something different.  I'm talking about putting of the other stuff in your hru loop into run_one_hru.

Reply:
Sort of. The routine `coupled_em` includes *most* of the physics for a single HRU. But I see what you mean. I can make a `run_oneHRU routine` that includes the HRU loop. It is indeed easier to do this now since I'm trimming the data structures anyway.

Reply:
As discussed we can leave the HRU break-out till later. In an MPI context there is an advantage anyway to use the GRU as the basis for domain decomposition, because there is potential exchange between HRUs within a GRU, but there is no exchange between GRUs. So we can move forward with MPI and later evaluate whether a further breakout is warranted.

Reply:
As discussed on slack - @martynpclark will address the simplification and HRU breakouts for subsequent PRs